identity-wallet/src/state/did/reducers/preferred_method.rs

```diff
@@ -0,0 +1,31 @@
+use crate::{
+    error::AppError,
+    state::{
+        actions::{listen, Action},
+        did::actions::set_preferred_method::SetPreferredDidMethod,
+        AppState,
+    },
+};
+use oid4vc::oid4vc_core::SubjectSyntaxType;
+use std::str::FromStr;
+
+pub async fn set_preferred_did_method(mut state: AppState, action: Action) -> Result<AppState, AppError> {
```

Comment:
I think it is important we keep the incoming state immutable, because of reasons described [here](https://github.com/impierce/identity-wallet/issues/163) and [here](https://github.com/impierce/identity-wallet/pull/157#discussion_r1526166365).

So this reducer can be rewritten as:
```rust
pub async fn set_preferred_did_method(state: AppState, action: Action) -> Result<AppState, AppError> {
    if let Some(method) = listen::<SetPreferredDidMethod>(action).map(|payload| payload.method) { // <-- This reducer listens only to the `SetPreferredDidMethod` Action, so it should build a new `AppState` inside this if statement.
        let mut managers = state.core_utils.managers.lock().await;

        let identity_manager = managers
            .identity_manager
            .as_mut()
            .ok_or(AppError::MissingManagerError("identity"))?;

        // TODO: add error handling here
        let subject_syntax_type = SubjectSyntaxType::from_str(&method.to_string()).unwrap();

        identity_manager.provider_manager.provider.default_subject_syntax_type = subject_syntax_type.clone();
        identity_manager.wallet.default_subject_syntax_type = subject_syntax_type;

        drop(managers);
        return Ok(AppState { // <-- Return a new 'fresh' `AppState` where it takes ownership over the fields from the previous state, and aadds updates to the state as well (in this case `preferred_did_method`).
            profile_settings: ProfileSettings {
                preferred_did_method: method.to_string(),
                ..state.profile_settings
            },
            current_user_prompt: None,
            ..state
        });
    }
    Ok(state) // <-- Outside the if statement it should not change the current state at all.
}
```

Reply:
You are right, I was carelessly mutating the state ... shame on me :'D